File: flutter_client/lib/utils/pos_qty_rules.dart
import 'package:intl/intl.dart';

import '../models/pos_product.dart';

/// Format / validate SL theo cờ [PosProduct.allowDecimalQty].
class PosQtyRules {
  PosQtyRules._();

  static final _decFmt = NumberFormat('#,##0.####', 'vi_VN');
  static final _intFmt = NumberFormat('#,##0', 'vi_VN');

  static bool isWhole(double qty) => qty == qty.roundToDouble();

  /// Hàng bắt buộc seri hoặc chưa bật thập phân → chỉ số nguyên.
  static bool mustBeWhole(PosProduct product) =>
      product.requiresSerial || !product.allowDecimalQty;

  static bool allowsDecimal(PosProduct product) =>
      product.allowDecimalQty && !product.requiresSerial;

  static String format(
    double qty, {
    PosProduct? product,
    bool? allowDecimal,
  }) {
    final dec = allowDecimal ??
        (product == null ? !isWhole(qty) : allowsDecimal(product));
    if (!dec || isWhole(qty)) return _intFmt.format(qty.round());
    return _decFmt.format(qty);
  }

  /// Null = hợp lệ.
  static String? validate(PosProduct product, double qty, {String action = 'Thao tác'}) {
    if (qty <= 0) return '$action: số lượng phải > 0 («${product.name}»).';
    if (mustBeWhole(product) && !isWhole(qty)) {
      if (product.requiresSerial) {
        return '$action: «${product.name}» bắt buộc seri — số lượng phải là số nguyên.';
      }
      return '$action: «${product.name}» chưa cho phép SL thập phân. Bật trong hàng hóa để dùng $qty.';
    }
    return null;
  }
}

File: src/ZKTecoADMS.Api/Controllers/PosQtyRules.cs
using ZKTecoADMS.Domain.Entities;

namespace ZKTecoADMS.Api.Controllers;

/// <summary>Quy tắc số lượng nguyên / thập phân theo cấu hình hàng hóa.</summary>
public static class PosQtyRules
{
    public static bool IsWhole(decimal qty) => qty == decimal.Truncate(qty);

    /// <summary>
    /// Null = hợp lệ. Không cho thập phân nếu hàng chưa bật AllowDecimalQty.
    /// Hàng bắt buộc seri luôn phải số nguyên.
    /// </summary>
    public static string? ValidateLineQty(PosProduct product, decimal qty, string actionLabel)
    {
        if (qty <= 0)
            return $"{actionLabel}: số lượng phải > 0 («{product.Name}»).";

        var mustBeWhole = product.RequiresSerial || !product.AllowDecimalQty;
        if (mustBeWhole && !IsWhole(qty))
        {
            if (product.RequiresSerial)
                return $"{actionLabel}: «{product.Name}» bắt buộc seri — số lượng phải là số nguyên.";
            return $"{actionLabel}: «{product.Name}» chưa cho phép SL thập phân. Bật trong hàng hóa để bán/nhập {qty}.";
        }

        return null;
    }
}
